# ĐỒ ÁN THỰC HÀNH

# A. Thu nhập dữ liệu
1. Chủ đề của dữ liệu là gì và được lấy từ nguồn nào?
- Tên chủ đề: Mức lương cho công việc Khoa học Dữ liệu (Data Science Job Salaries).
- Nguồn dữ liệu: [tại đây](https://salaries.ai-jobs.net/)
2. Tác giả có cho phép sử dụng dữ liệu này không?
- Toàn bộ bộ dữ liệu được công bố trong phạm vi công cộng theo giấy phép CC0 (Public Domain).
3. Tác giả thu thập dữ liệu như thế nào?
- Trang web cung cấp cho người dùng một biểu mẫu ngắn để trả lời khảo sát về công việc và mức lương hiện tại trong lĩnh vực Khoa học dữ liệu

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_rows", 500)
pd.set_option("display.max_columns", 500)

# B. Khám phá dữ liệu

## 1. Đọc dữ liệu

In [2]:
df = pd.read_csv("./data/salaries.csv")
df.head()

work_year experience_level employment_type             job_title  salary  \
0       2023               SE              FT          Data Modeler  258700   
1       2023               SE              FT          Data Modeler  146600   
2       2023               MI              FT  Data Science Manager  190000   
3       2023               MI              FT  Data Science Manager  160000   
4       2023               EN              FT      Research Analyst   90000   

  salary_currency  salary_in_usd employee_residence  remote_ratio  \
0             USD         258700                 US             0   
1             USD         146600                 US             0   
2             USD         190000                 US           100   
3             USD         160000                 US           100   
4             USD          90000                 US           100   

  company_location company_size  
0               US            M  
1               US            M  
2               US            M  
3               US            M  
4               US            M

## 2. Kích thước dữ liệu

In [3]:
num_rows, num_cols = df.shape
print("Số dòng:", num_rows)
print("Số cột:", num_cols)

Số dòng: 10456
Số cột: 11


## 3. Ý nghĩa từng dòng dữ liệu
- Mỗi dòng dữ liệu là thông tin về công ty và công việc của người khảo sát.

## 4. Dữ liệu có các dòng bị lặp không?

In [4]:
duplicate_rows = df.loc[df.duplicated()]

if duplicate_rows.shape[0] == 0: 
    print("Dữ liệu không có các dòng bị lặp")
else: 
    print("Dữ liệu các dòng bị lặp")
    print("Số dòng bị lặp là:", duplicate_rows.shape[0])

Dữ liệu các dòng bị lặp
Số dòng bị lặp là: 4428


## 5. Ý nghĩa từng cột dữ liệu

In [5]:
print(f"Dữ liệu có {len(df.columns)} cột")
print("Ý nghĩa các cột:\n", list(df.columns))

Dữ liệu có 11 cột
Ý nghĩa các cột:
 ['work_year', 'experience_level', 'employment_type', 'job_title', 'salary', 'salary_currency', 'salary_in_usd', 'employee_residence', 'remote_ratio', 'company_location', 'company_size']


## 6. Kiểu dữ liệu của mỗi cột dữ liệu. Có cột nào có kiểu dữ liệu không phù hợp không?

In [6]:
df.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

- Các thuộc tính của tập dữ liệu đa phần đã ở định dạng phù hợp.

## 7. Với mỗi cột mang giá trị số, các giá trị được phân bố như thế nào?

In [7]:
numerical_df = df.select_dtypes(include=["int64", "float64"])
numerical_df.head()

work_year  salary  salary_in_usd  remote_ratio
0       2023  258700         258700             0
1       2023  146600         146600             0
2       2023  190000         190000           100
3       2023  160000         160000           100
4       2023   90000          90000           100

### 7.1. Phần trăm giá trị còn thiếu là bao nhiêu?

In [8]:
numerical_df_null = (numerical_df.isnull().sum() / len(df)) * 100

print("Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính số:")
print(numerical_df_null)

Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính số:
work_year        0.0
salary           0.0
salary_in_usd    0.0
remote_ratio     0.0
dtype: float64


### 7.2. Giá trị nhỏ nhất? Giá trị lớn nhất? Các giá trị này có gì bất thường không?

In [9]:
numerical_df.describe()

work_year        salary  salary_in_usd  remote_ratio
count  10456.000000  1.045600e+04   10456.000000   10456.00000
mean    2022.778883  1.704655e+05  149263.499904      36.49101
std        0.506937  4.195989e+05   65269.393738      47.58074
min     2020.000000  1.400000e+04   15000.000000       0.00000
25%     2023.000000  1.050000e+05  104640.250000       0.00000
50%     2023.000000  1.430000e+05  142200.000000       0.00000
75%     2023.000000  1.882750e+05  185900.000000     100.00000
max     2023.000000  3.040000e+07  750000.000000     100.00000

- Các giá trị không có gì là bất thường.

## 8. Với mỗi cột mang giá trị phân loại, các giá trị được phân bố như thế nào?

In [10]:
categorical_df = df.select_dtypes(exclude=["int64", "float64"])
categorical_df.head()

experience_level employment_type             job_title salary_currency  \
0               SE              FT          Data Modeler             USD   
1               SE              FT          Data Modeler             USD   
2               MI              FT  Data Science Manager             USD   
3               MI              FT  Data Science Manager             USD   
4               EN              FT      Research Analyst             USD   

  employee_residence company_location company_size  
0                 US               US            M  
1                 US               US            M  
2                 US               US            M  
3                 US               US            M  
4                 US               US            M

### 8.1. Phần trăm giá trị còn thiếu là bao nhiêu?

In [11]:
categorical_df_null = (categorical_df.isnull().sum() / len(df)) * 100

print("Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính phân loại:")
print(categorical_df_null)

Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính phân loại:
experience_level      0.0
employment_type       0.0
job_title             0.0
salary_currency       0.0
employee_residence    0.0
company_location      0.0
company_size          0.0
dtype: float64


### 8.2. Mỗi cột có bao nhiêu giá trị khác nhau?

In [12]:
for column in categorical_df.columns:
    unique_values = categorical_df[column].unique()
    print(f"+ Cột {column}: {unique_values}")

+ Cột experience_level: ['SE' 'MI' 'EN' 'EX']
+ Cột employment_type: ['FT' 'PT' 'CT' 'FL']
+ Cột job_title: ['Data Modeler' 'Data Science Manager' 'Research Analyst' 'Data Engineer'
 'Data Analyst' 'ETL Developer' 'Machine Learning Engineer'
 'Data Scientist' 'Data Specialist' 'Data Strategist' 'Prompt Engineer'
 'Business Intelligence Developer' 'Research Engineer'
 'Business Intelligence Analyst' 'Data Science Lead' 'ML Engineer'
 'Analytics Engineer' 'BI Developer' 'Data Quality Manager'
 'Applied Scientist' 'Head of Data' 'Research Scientist' 'Data Architect'
 'Data Science' 'Business Intelligence Engineer' 'Data Science Consultant'
 'Machine Learning Scientist' 'Data Product Manager'
 'Business Intelligence Manager' 'Data Manager' 'Data Analytics Manager'
 'Computer Vision Engineer' 'AI Product Manager' 'Data Analytics Lead'
 'Director of Data Science' 'Data Product Owner' 'Data Developer'
 'Machine Learning Infrastructure Engineer' 'Data Lead' 'AI Engineer'
 'Data Integration Eng

### 8.3. Các giá trị này có gì bất thường không?
- Các giá trị không có gì là bất thường. Tuy nhiên trong cột `nghề nghiệp (job_title)` có quá nhiều nghề, ta có thể rút gọn các nghề này vào một lĩnh vực.

# C. Đưa ra câu hỏi có ý nghĩa và trả lời câu hỏi

## Câu hỏi 5: Quốc gia nào cung cấp nguồn thu nhập tốt nhất cho các chuyên gia trong lĩnh vực Khoa học Dữ liệu khi họ lựa chọn ngành nghề?
- Tiền xử lý:
    - Do `vị trí công ty (company_location)` đang ở dạng mã ISO 3166-1 alpha-2 nên ta sẽ chuyển sang tên quốc gia dựa trên dạng mã đó.
    - Bảng mã ISO 3166-1 alpha-2 ta sẽ lấy trên Wikipedia [tại đây](https://vi.wikipedia.org/wiki/ISO_3166-1_alpha-2).
    - Sau đó ta sẽ kết hợp 2 bộ dữ liệu lại và đặt tên dữ liệu mới là `df_iso`.
- Phân tích dữ liệu:
    - Ta nhóm dữ liệu dựa trên `tên quốc gia (country_name)` và tính mức lương trung bình dựa trên `mức lương (salary_in_usd)`.
    - Sau đó ta làm tròn mức lương trung bình lên 2 chữ số thập phân và sắp xếp dữ liệu tăng dần.
    - Ta cũng sẽ xem qua mức lương ở Việt Nam.

In [13]:
iso = pd.read_html("https://vi.wikipedia.org/wiki/ISO_3166-1_alpha-2", match="Alpha-2")[0]
iso.rename(columns={"Tên quốc gia": "country_name"}, inplace=True)
iso.drop(["Ghi chú"], axis=1, inplace=True)

df_iso = pd.merge(df, iso, left_on="company_location", right_on="Alpha-2", how="left")

average_salary_by_country = df_iso.groupby("country_name")["salary_in_usd"].mean().reset_index()
average_salary_by_country.rename(columns={"salary_in_usd": "mean_salary_usd"}, inplace=True)
average_salary_by_country["mean_salary_usd"] = average_salary_by_country["mean_salary_usd"].round(2)
average_salary_by_country = average_salary_by_country.sort_values(by="mean_salary_usd", ascending=False).reset_index(drop=True)
average_salary_by_country

country_name  mean_salary_usd
0                       Qatar        300000.00
1                      Israel        217332.00
2                 Puerto Rico        167500.00
3               United States        157536.04
4                 New Zealand        151634.33
5                      Canada        144004.46
6                 Ả Rập Saudi        134999.00
7                   Australia        127624.77
8                     Ukraina        121333.33
9      Bosnia and Herzegovina        120000.00
10                      Japan        110821.62
11                    Ireland        110236.88
12             United Kingdom        100791.88
13                    Algérie        100000.00
14                       Iraq        100000.00
15  Iran, Islamic Republic of        100000.00
16                  Mauritius        100000.00
17       United Arab Emirates        100000.00
18                      China        100000.00
19                  Thụy Điển         98791.33
20                      Litva         97611.00
21                     México         94864.64
22                        Đức         92954.66
23                    Thụy Sĩ         88769.83
24                      Na Uy         88462.00
25                       Pháp         80337.76
26                  Gibraltar         79976.00
27                   Colombia         79115.25
28         Russian Federation         78207.86
29                     Hà Lan         76968.90
30                         Bỉ         76864.75
31                    Croatia         76726.00
32                         Áo         71354.83
33             Czech Republic         69478.67
34                   Phần Lan         68519.33
35                  Hong Kong         65058.00
36                      Kenya         65000.00
37                   Viet Nam         63000.00
38               South Africa         62935.40
39                  Singapore         62783.00
40                     Poland         62139.00
41                  Argentina         62000.00
42                    Nigeria         60444.44
43                     Latvia         58626.33
44                     Brasil         58569.12
45                Philippines         57670.00
46                   Slovenia         56186.33
47                Tây Ban Nha         55825.36
48                     Hy Lạp         52260.86
49                    România         50950.67
50                    Andorra         50745.00
51                    Armenia         50000.00
52                          Ý         49997.08
53                 Bồ Đào Nha         49787.08
54                   Đan Mạch         49402.60
55   Central African Republic         49216.00
56                 Luxembourg         47609.33
57         Korea, Republic of         47000.00
58                    Estonia         46706.10
59                    Bahamas         45555.00
60                      India         40287.11
61                      Chile         40038.00
62                   Malaysia         40000.00
63                  Indonesia         34208.00
64             American Samoa         34026.50
65                    Hungary         32139.67
66                   Pakistan         30000.00
67                      Malta         28369.00
68                      Ghana         27000.00
69                   Thailand         22971.33
70                     Ai Cập         22800.00
71                     Turkey         22313.60
72                   Honduras         20000.00
73       Moldova, Republic of         18000.00
74                    Ecuador         16000.00

In [14]:
vn_salary = average_salary_by_country.loc[average_salary_by_country["country_name"] == "Viet Nam"]
salary = vn_salary["mean_salary_usd"].values[0]
index = vn_salary["mean_salary_usd"].index[0] + 1 
print(f"Mức thu nhập trung bình hàng năm của Việt Nam là {salary} và đứng thứ {index} thế giới.")

Mức thu nhập trung bình hàng năm của Việt Nam là 63000.0 và đứng thứ 38 thế giới.


- Nhận xét:
    - Tổng quan:
        - Quốc gia trả lương cao nhất đó là Qatar với mức lương trung bình là 300000 USD và Israel với mức lương trung bình là	217332 USD. Đây là các quốc gia đều nằm ở Trung Đông.
        - Quốc gia trả lương thấp nhất nhất đó là Ecuador và mức lương trung bình là 16000.
        - Trong top 10 quốc gia trả lương cao nhất, có đến 5 quốc gia có ngôn ngữ chính thức là tiếng Anh: Puerto Rico, United States, New Zealand, Canada, Australia.
    - Việt Nam:
        - Mức thu nhập trung bình hàng năm của Việt Nam là 63000 USD (~ 1.5 tỷ VND) và đứng thứ 38 thế giới. Đây là một mức lương khá cao nếu so với mức thu nhập ở Việt Nam.
        - Khi xem xét lại dữ liệu, để đạt mức lương như trên, thì đa số đều là những người ở trình độ Mid-level / Intermediate, thời gian làm việc là Full-time và làm trong công ty vừa và lớn trở lên.